### ISSUE TITLE
failed to resolve: could not find `implementation` in the crate root When use actor_interface

### ISSUE THREAD

**Issue by YeahhhhzZ** [s2e-systems/dust-dds#259]
dust_dds_derive v0.8.2.
https://docs.rs/dust_dds_derive/0.8.2/dust_dds_derive/attr.actor_interface.html

```rust
use dust_dds::{
    dds_async::{topic::TopicAsync, topic_listener::TopicListenerAsync},
    infrastructure::status::InconsistentTopicStatus,
};
use dust_dds_derive::actor_interface;
pub struct TopicListenerActor {
    listener: Option<Box<dyn TopicListenerAsync + Send>>,
}

#[actor_interface]
impl TopicListenerActor {
    async fn on_inconsistent_topic(
        &mut self,
        the_topic: TopicAsync,
        status: InconsistentTopicStatus,
    ) {
        if let Some(l) = &mut self.listener {
            l.on_inconsistent_topic(the_topic, status).await
        }
    }
}

fn main() {
    println!("Hello, world!");
}
```

```txt
$ cargo run
   Compiling dust_dds_examples v0.1.0 
error[E0433]: failed to resolve: could not find `implementation` in the crate root
  --> src/main.rs:10:1
   |
10 | #[actor_interface]
   | ^^^^^^^^^^^^^^^^^^ could not find `implementation` in the crate root
   |
   = note: this error originates in the attribute macro `actor_interface` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0433`.
error: could not compile `dust_dds_examples` (bin "dust_dds_examples") due to previous error
```

**Comment by jrebelo** [comment]
Hi @YeahhhhzZ,  the #[actor_interface] macro is not meant to be used outside of the dust_dds implementation crate. As a user of dust_dds you should only depend on the derive functions which are re-exported by dust_dds instead of directly using the dust_dds_derive crate. For reference all those derives are for the traits in [type_support](https://docs.rs/dust_dds/latest/dust_dds/topic_definition/type_support/index.html).

If I understand correctly you are trying to install a listener on the topic. To do this you first need to create the object which will be the listener and implement the TopicListenerAsync trait on it. Here is an example on how to do that:
```rust
struct ExampleTopicListener {
    inconsistent_count: i32,
}

impl dust_dds::dds_async::topic_listener::TopicListenerAsync for ExampleTopicListener {
    fn on_inconsistent_topic(
        &mut self,
        _the_topic: dust_dds::dds_async::topic::TopicAsync,
        _status: dust_dds::infrastructure::status::InconsistentTopicStatus,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = ()> + Send + '_>> {
        Box::pin(async {
            self.inconsistent_count += 1;
        })
    }
}
```

Once you have such object you can install it when creating the topic by passing the optional box and the status kind you want it to react on:
```rust
let topic = participant
        .create_topic::<UserData>(
            "LargeDataTopic",
            "UserData",
            QosKind::Default,
            Some(Box::new(ExampleTopicListener {
                inconsistent_count: 0,
            })),
            &[StatusKind::InconsistentTopic],
        )
        .await
        .unwrap();
```

**Comment by YeahhhhzZ** [comment]
> Hi @YeahhhhzZ, the #[actor_interface] macro is not meant to be used outside of the dust_dds implementation crate. As a user of dust_dds you should only depend on the derive functions which are re-exported by dust_dds instead of directly using the dust_dds_derive crate. For reference all those derives are for the traits in [type_support](https://docs.rs/dust_dds/latest/dust_dds/topic_definition/type_support/index.html).
> 
> If I understand correctly you are trying to install a listener on the topic. To do this you first need to create the object which will be the listener and implement the TopicListenerAsync trait on it. Here is an example on how to do that:
> 
> ```rust
> struct ExampleTopicListener {
>     inconsistent_count: i32,
> }
> 
> impl dust_dds::dds_async::topic_listener::TopicListenerAsync for ExampleTopicListener {
>     fn on_inconsistent_topic(
>         &mut self,
>         _the_topic: dust_dds::dds_async::topic::TopicAsync,
>         _status: dust_dds::infrastructure::status::InconsistentTopicStatus,
>     ) -> std::pin::Pin<Box<dyn std::future::Future<Output = ()> + Send + '_>> {
>         Box::pin(async {
>             self.inconsistent_count += 1;
>         })
>     }
> }
> ```
> 
> Once you have such object you can install it when creating the topic by passing the optional box and the status kind you want it to react on:
> 
> ```rust
> let topic = participant
>         .create_topic::<UserData>(
>             "LargeDataTopic",
>             "UserData",
>             QosKind::Default,
>             Some(Box::new(ExampleTopicListener {
>                 inconsistent_count: 0,
>             })),
>             &[StatusKind::InconsistentTopic],
>         )
>         .await
>         .unwrap();
> ```

Thanks, I originally want to package different listeners into a structure, like:
```rust
struct TopicListener {
    listener: xxx,
    mask: Vec<StatusKind>,
};
// other Listener struct

// package
struct Listeners {
    topic: Option<TopicListener>,
    publisher:  Option<PublisherListener>,
    subscriber: Option<SubscriberListener>,
    datawriter: Option<DataWriterListener>,
    datareader: Option<DataReaderListener>,
}
```
and provide it to the user so that the user can customize the listener as needed, and then provide a create_writer/create_reader method, like:
```rust
fn create_datawriter(topic_name: &str, listeners: Listeners) -> DDSWriter {}
fn create_datareader(topic_name: &str, listeners: Listeners) -> DDSReader {}
```
Later, I saw the related usage of actor_interface in the dust_dds_derive document and thought about giving it a try.

Now, let me try again according to your method

**Comment by jrebelo** [comment]
Yes, you can indeed implement the multiple listener traits on the same object and use different instances of that object as a listener for different entities. I will close this issue for now as this is working as intended from a Dust DDS perspective.
